main.go

```diff
@@ -145,3 +194,76 @@ func loadConfig(cfgJSON *extapi.JSON) (customDNSProviderConfig, error) {
 
 	return cfg, nil
 }
+
+func (c *ns1DNSProviderSolver) setNS1Client(ch *v1alpha1.ChallengeRequest, cfg ns1DNSProviderConfig) error {
+	ref := cfg.APIKeySecretRef
+	if ref.Key == "" {
+		return fmt.Errorf(
+			"no APIKey for %q in secret '%s/%s'",
+			ref.Name,
+			ref.Key,
+			ch.ResourceNamespace,
+		)
+	}
+	if ref.Name == "" {
+		return fmt.Errorf(
+			"no APIKey for %q in secret '%s/%s'",
+			ref.Name,
+			ref.Key,
+			ch.ResourceNamespace,
+		)
+	}
+
+	secret, err := c.k8sClient.CoreV1().Secrets(ch.ResourceNamespace).Get(
+		ref.Name, metav1.GetOptions{},
+	)
+	if err != nil {
+		return err
+	}
+	apiKeyBytes, ok := secret.Data[ref.Key]
+	if !ok {
+		return fmt.Errorf(
+			"no APIKey for %q in secret '%s/%s'",
+			ref.Name,
+			ref.Key,
+			ch.ResourceNamespace,
+		)
+	}
+	apiKey := string(apiKeyBytes)
+
+	httpClient := &http.Client{}
+	if cfg.IgnoreSSL == true {
+		tr := &http.Transport{
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+		}
+		httpClient.Transport = tr
+	}
+	c.ns1Client = ns1API.NewClient(
+		httpClient,
+		ns1API.SetAPIKey(apiKey),
+		ns1API.SetEndpoint(cfg.Endpoint),
+	)
+
+	return nil
+}
+
+// Get the zone and domain we are setting from the challenge request
+func (c *ns1DNSProviderSolver) parseChallenge(ch *v1alpha1.ChallengeRequest) (
+	zone string, domain string, err error,
+) {
+
+	zone, err = util.FindZoneByFqdn(ch.ResolvedFQDN, util.RecursiveNameservers)
+	if err == nil {
```

Comment:
this should check for nil error and return, drop the else statement and outdent the `zone = ...`